File: okj/szoftverfejleszto/acb_kosarlabdaliga/python/Kosar2004.py
from datetime import date
from collections import Counter

class Eredmeny:

    def __init__(self, line: str):
        split = line.split(';')

        self.hazaiCsapat = split[0]
        self.idegenCsapat = split[1]
        self.hazaiPont = int(split[2])
        self.idegenPont = int(split[3])
        self.helyszin = split[4]
        self.idopont = date.fromisoformat(split[5].strip())

with open('eredmenyek.csv') as file:
    lines = file.readlines()
    eredmenyek = [ Eredmeny(lines[i]) for i in range(1, len(lines)) ]

hazaiMadridDb  = sum(1 for k in eredmenyek if k.hazaiCsapat  == 'Real Madrid')
idegenMadridDb = sum(1 for k in eredmenyek if k.idegenCsapat == 'Real Madrid')

print(f'3. Feladat: Hazai: {hazaiMadridDb}, idegen: {idegenMadridDb}')

voltEDontetlen = any(k for k in eredmenyek if k.hazaiPont == k.idegenPont)
print(f'4. Feladat: Volt e döntetlen: {"igen" if voltEDontetlen else "nem"}')

barcelona = next(k for k in eredmenyek if 'Barcelona' in k.hazaiCsapat or 'Barcelona' in k.idegenCsapat)
print(f'5. Feladat: Barcelona csapat neve: {barcelona.hazaiCsapat if "Barcelona" in barcelona.hazaiCsapat else barcelona.idegenCsapat}')
print('6. Feladat:')

hatosFeladatIdopont = date(2004, 11, 21)
for eredmeny in eredmenyek:
    if eredmeny.idopont == hatosFeladatIdopont:
        print(f'    {eredmeny.hazaiCsapat} - {eredmeny.idegenCsapat} ({eredmeny.hazaiPont}:{eredmeny.idegenPont})')

print('7. Feladat:')

stat = Counter(k.helyszin for k in eredmenyek)
for helyszin, db in stat.items():
    if db > 20:
        print(f'    {helyszin}: {db}')

File: okj/szoftverfejleszto/acb_kosarlabdaliga/rust/Kosar2004.rs
use std::{ fs::File, io::BufReader, io::BufRead };
use encoding_rs::WINDOWS_1252;
use encoding_rs_io::DecodeReaderBytesBuilder;
use chrono::NaiveDate;
use itertools::Itertools;

fn main() {
    let input_file = File::open("eredmenyek.csv").unwrap();
    let eredmenyek = BufReader::new(DecodeReaderBytesBuilder::new().encoding(Some(WINDOWS_1252)).build(input_file))
                              .lines()
                              .skip(1)
                              .map(|k| create_eredmeny(&k.unwrap()))
                              .collect::<Vec<_>>();

    let hazai_madrid_db = eredmenyek.iter()
                                    .filter(|k| k.hazai_csapat == "Real Madrid")
                                    .count();

    let idegen_madrid_db = eredmenyek.iter()
                                     .filter(|k| k.idegen_csapat == "Real Madrid")
                                     .count();

    println!("3. Feladat: Hazai: {}, idegen: {}", hazai_madrid_db, idegen_madrid_db);

    match eredmenyek.iter().any(|k| k.hazai_pont == k.idegen_pont) {
        false => println!("4. Feladat: Nem volt döntetlen"),
        true => println!("4. Feladat: Volt döntetlen")
    };

    match eredmenyek.iter().find(|k| k.hazai_csapat.contains("Barcelona") || k.idegen_csapat.contains("Barcelona")) {
        None => {},
        Some(k) => println!("5. Feladat: Barcelona csapat neve: {}", if k.hazai_csapat.contains("Barcelona") { &k.hazai_csapat } else { &k.idegen_csapat })
    }

    let hatos_feladat_idopont = NaiveDate::from_ymd_opt(2004, 11, 21).unwrap();

    println!("6. Feladat:");

    eredmenyek.iter()
              .filter(|k| k.idopont == hatos_feladat_idopont)
              .for_each(|k| println!("    {} - {} ({}:{})", k.hazai_csapat, k.idegen_csapat, k.hazai_pont, k.idegen_pont));

    println!("7. Feladat:");

    eredmenyek.iter()
              .sorted_by_key(|k| &k.helyszin)
              .group_by(|k| &k.helyszin)
              .into_iter()
              .map(|(helyszin, eredmenyek_helyszinhez)| (helyszin, eredmenyek_helyszinhez.count()))
              .filter(|(_, db)| *db > 20)
              .for_each(|(helyszin, db)| println!("    {}: {}", helyszin, db));
}

fn create_eredmeny(line: &String) -> Eredmeny {
    let split = line.split(';').collect::<Vec<_>>();

    Eredmeny {
        hazai_csapat: split[0].to_string(),
        idegen_csapat: split[1].to_string(),
        hazai_pont: split[2].parse::<i32>().unwrap(),
        idegen_pont: split[3].parse::<i32>().unwrap(),
        helyszin: split[4].to_string(),
        idopont: NaiveDate::parse_from_str(split[5], "%Y-%m-%d").unwrap()
    }
}

struct Eredmeny {
    hazai_csapat: String,
    idegen_csapat: String,
    hazai_pont: i32,
    idegen_pont: i32,
    helyszin: String,
    idopont: NaiveDate
}